bson/src/main/org/bson/codecs/pojo/PojoBuilderHelper.java

```diff
@@ -149,15 +157,30 @@ private static TypeParameterMap getTypeParameterMap(final List<String> genericTy
     }
 
     @SuppressWarnings("unchecked")
-    private static <T> TypeData<T> getFieldTypeDataFromClass(final Class<T> type) {
-        TypeData.Builder<T> builder = TypeData.builder(type);
-        if (Collection.class.isAssignableFrom(type)) {
-            builder.addTypeParameter(TypeData.builder(Object.class).build());
-        } else if (Map.class.isAssignableFrom(type)) {
-            builder.addTypeParameter(TypeData.builder(String.class).build());
-            builder.addTypeParameter(TypeData.builder(Object.class).build());
+    private static <V> void specializeFieldModelBuilder(final FieldModelBuilder<V> fieldModelBuilder,
+                                                        final TypeParameterMap typeParameterMap,
+                                                        final List<TypeData<?>> fieldTypeParameters) {
+        if (typeParameterMap.hasTypeParameters() && !fieldTypeParameters.isEmpty()) {
+            TypeData<V> specializedFieldType = fieldModelBuilder.getTypeData();
+            Map<Integer, Integer> fieldToClassParamIndexMap = typeParameterMap.getFieldToClassParamIndexMap();
+            Integer classTypeParamRepresentsWholeField = fieldToClassParamIndexMap.get(-1);
+            if (classTypeParamRepresentsWholeField != null) {
+                specializedFieldType = (TypeData<V>) fieldTypeParameters.get(classTypeParamRepresentsWholeField);
```

Comment:
I'm not clear on why the generic support worked before without this line of code.   Can you clarify?